plugins/image-prioritizer/class-image-prioritizer-img-tag-visitor.php

```diff
@@ -28,7 +42,37 @@ final class Image_Prioritizer_Img_Tag_Visitor extends Image_Prioritizer_Tag_Visi
 	 */
 	public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 		$processor = $context->processor;
-		if ( 'IMG' !== $processor->get_tag() ) {
+		$tag       = $processor->get_tag();
+
+		// Handle opening and closing `<picture>` tags.
+		if ( 'PICTURE' === $tag ) {
+			if ( ! $processor->is_tag_closer() ) {
+				// Opening <picture> tag.
+				$this->is_within_picture = true;
+				$this->collected_sources = array();
+			} else {
+				// Closing </picture> tag.
+				$this->is_within_picture = false;
+				$this->collected_sources = array();
+			}
+
+			return false;
+		}
+
+		// Collect `<source>` elements within `<picture>`.
+		if ( $this->is_within_picture && 'SOURCE' === $tag && ! $processor->is_tag_closer() ) {
+			$this->collected_sources[] = array(
+				'srcset'      => $processor->get_attribute( 'srcset' ),
+				'sizes'       => $processor->get_attribute( 'sizes' ),
+				'type'        => $processor->get_attribute( 'type' ),
+				'media'       => $processor->get_attribute( 'media' ),
+				'crossorigin' => $this->get_attribute_value( $processor, 'crossorigin' ),
+			);
```

Comment:
I think this needs some additional logic. We need to prevent a scenario where there is a `PICTURE` with the first `SOURCE` being an AVIF image and then the second being a WebP, with the third being a JPEG fallback. We must only ever add a preload link for the first `SOURCE` because otherwise if a browser supports AVIF and WebP then they could end up preloading both but only the first would be used.

Additionalyl, 

See comments:

* https://github.com/WordPress/performance/issues/1312#issuecomment-2189266083
* https://github.com/WordPress/performance/issues/1312#issuecomment-2189958858

Especially when there is no `media` attribute but only `type`, it's important that only a preload link for the first `SOURCE` be added. This is because

Reply:
Oops. Sorry for the poor comment. Not sure how that happened.

> 1. Picture element with multiple `source` tags without `media` attribute.

And naturally _with_ `type` attributes.

> 2\. Picture element with multiple `source` tags with `media` attribute.

And _without_ `type` attributes.

> 3\. Picture element with no `source` tags.

This would be an error scenario. A PICTURE tag without a SOURCE is not valid. So just do nothing in this case.

> 4\. Picture element with multiple `source` but some have `media` attribute and some don't have `media` attribute.

I think this would be an unusual case. I suggest we skip this scenario. 

In fact, I suggest we may want to skip handling PICTURE tags that have SOURCE tags with `media` attributes entirely. This greatly complicate things, since we'd need to make sure that only one of the SOURCE tags ever get preloaded, but this would require computing some negated media query or something. As I recall, when a PICTURE decides which SOURCE to use, it will iterate over the SOURCE tags until it arrives at the first one that has a matching `media` and has a supported `type`. But this first-match logic is difficult to replicate for preload links.

Since our Modern Image Formats plugin only outputs SOURCE tags with `type` attributes, this would be the most important to support.

We can leave support for `media` attributes for a future enhancement, especially as we find there is a need for it.

Reply:
Gemini concurs that preloading SOURCE with media attributes is not feasible, after initially getting it wrong: https://g.co/gemini/share/57ac56545028

So yeah, for PICTURE, grab the first SOURCE child and only preload it if none of the SOURCE tags have a `media` attribute. The only way to properly prioritize such an image is to ensure that `fetchpriority=high` is on the IMG tag, which we should make sure is happening in addition to adding preload links.

Reply:
I have now implemented this in [df1d90e](https://github.com/b1ink0/performance/commit/df1d90e91e032c5544a083a704b843b78fa920fc)
> So yeah, for PICTURE, grab the first SOURCE child and only preload it if none of the SOURCE tags have a media attribute. 

But I have a doubt about the part
> The only way to properly prioritize such an image is to ensure that `fetchpriority=high` is on the IMG tag, which we should make sure is happening in addition to adding preload links.

Currently preload link has `fetchpriority=high` set but the image of that picture element has `data-od-removed-fetchpriority="high"` should we add `fetchpriority=high` to the image and remove the added `data-od-removed-fetchpriority` aatribute?

Reply:
OK, I just tested this out and it is actually working currently as expected. Chrome is reporting that the LCP element is the `PICTURE > IMG` tag, _not_ the `PICTURE` wrapper tag:

![image](https://github.com/user-attachments/assets/6c555e8c-b9a9-4c18-ae80-a265920a9bf5)

So when all viewport groups are populated with URL Metrics, it is resulting in the following markup in `trunk`:

```html
<picture class="wp-picture-394" style="display: contents"
  ><source
    type="image/webp"
    srcset="
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-1024x683-jpg.webp  1024w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-300x200-jpg.webp    300w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-768x512-jpg.webp    768w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-1536x1024-jpg.webp 1536w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-jpg.webp           1620w
    "
    sizes="(max-width: 1024px) 100vw, 1024px" />
  <img
    data-od-fetchpriority-already-added=""
    data-od-xpath="/*[1][self::HTML]/*[2][self::BODY]/*[3][self::DIV]/*[3][self::DIV]/*[1][self::DIV]/*[1][self::MAIN]/*[1][self::ARTICLE]/*[2][self::DIV]/*[1][self::FIGURE]/*[1][self::PICTURE]/*[2][self::IMG]"
    data-dominant-color="8a9464"
    data-has-transparency="false"
    style="--dominant-color: #8a9464"
    fetchpriority="high"
    decoding="async"
    width="1024"
    height="683"
    src="http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-1024x683.jpg"
    alt=""
    class="wp-image-394 not-transparent"
    srcset="
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-1024x683.jpg  1024w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-300x200.jpg    300w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-768x512.jpg    768w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-1536x1024.jpg 1536w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi-1568x1045.jpg 1568w,
      http://localhost:8888/wp-content/uploads/2024/11/1620px-Australasian_swamphen_Porphyrio_melanotus_Tiritiri_Matangi.jpg           1620w
    "
    sizes="(max-width: 1024px) 100vw, 1024px"
/></picture>
```

I believe you saw `data-od-removed-fetchpriority="high"` being added because you didn't have URL Metrics collected for some of the viewport groups, and currently the [`OD_URL_Metric_Group_Collection::get_common_lcp_element()`](https://github.com/WordPress/performance/blob/69eca9a7936701475f84a0cabbfec36e1cb3fe27/plugins/optimization-detective/class-od-url-metric-group-collection.php#L392-L442) method is defined so that _all_ URL Metrics groups must be populated in order for the method to return an element. I think this should be changed to align better with how we determine whether to do lazy-loading: if both the narrowest viewport group (i.e. mobile) and the widest viewport group (i.e. desktop) are populated with URL Metrics and these groups both return an LCP Element via `$group->get_lcp_element()` which has the same XPath, then we should consider that it is likely for all of the viewport groups to have the same LCP element (i.e. phablet and tablet).

Reply:
I've filed this in https://github.com/WordPress/performance/issues/1710.